File: Python/Medium/graphs/KeysandRooms.py
import collections
class Solution:
    def canVisitAllRooms(self, rooms):
        graph=collections.defaultdict(lambda: [])
        n=len(rooms)
        keys=set()
        for i in range(n):
            graph[i]=rooms[i]
            if rooms[i]==i:
                return False

        def DFS(key):
            if key in keys: #may not be right
                return 
            
            keys.add(key)
            for neighbor in graph[key]:
                DFS(neighbor)
        
        DFS(0)
        if list(keys)==[i for i in range(n)]:
            return True
        else:
            return False

rooms = [[1],[2],[3],[]]
rooms2 = [[1,3],[3,0,1],[2],[0]]
print(Solution().canVisitAllRooms(rooms))
            
        

        

File: JavaScript/Medium/graphs/keysAndRooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
var canVisitAllRooms = function(rooms) {
    let graph = new Array(rooms.length).fill(0)
    for (let i = 0; i<graph.length; i++){
        graph[i]=rooms[i]
        if (rooms[i]===i) {return false}
    }

    let keys = new Set()

    var DFS = (key) => {
        if (keys.has(key)){return}

        keys.add(key)
        for (const neighbor of graph[key]){
            DFS(neighbor)
        }
    }
    var sortHelper = (a,b) =>{return a-b}

    DFS(0)

    if (rooms.length==keys.size) {
        return true
    } else {
        return false}

    
};

rooms = [[1],[2],[3],[]]
console.log(canVisitAllRooms(rooms))